engine/table/src/main/java/io/deephaven/engine/table/impl/locations/local/FileKeyValuePartitionLayout.java

```diff
@@ -0,0 +1,152 @@
+     *        look at files in {@code tableRootDirectory} and find no partitions.
+     */
+    public FileKeyValuePartitionLayout(
+            @NotNull final File tableRootDirectory,
+            @NotNull final Predicate<Path> pathFilter,
+            @NotNull final Supplier<LocationTableBuilder> locationTableBuilderFactory,
+            @NotNull final BiFunction<Path, Map<String, Comparable<?>>, TLK> keyFactory,
+            final int maxPartitioningLevels) {
+        super(keyFactory);
+        this.tableRootDirectory = tableRootDirectory;
+        this.pathFilter = pathFilter;
+        this.locationTableBuilderFactory = locationTableBuilderFactory;
+        this.maxPartitioningLevels = Require.geqZero(maxPartitioningLevels, "maxPartitioningLevels");
+    }
+
+    public String toString() {
+        return FileKeyValuePartitionLayout.class.getSimpleName() + '[' + tableRootDirectory + ']';
+    }
+
+    @Override
+    public void findKeys(@NotNull final Consumer<TLK> locationKeyObserver) {
+        final Deque<Path> targetFiles = new ArrayDeque<>();
+        final LocationTableBuilder locationTableBuilder = locationTableBuilderFactory.get();
+        try {
+            Files.walkFileTree(tableRootDirectory.toPath(), EnumSet.of(FileVisitOption.FOLLOW_LINKS),
+                    maxPartitioningLevels + 1, new SimpleFileVisitor<>() {
+                        final List<String> partitionKeys = new ArrayList<>();
+                        final List<String> partitionValues = new ArrayList<>();
+                        boolean registered;
+                        int columnCount = -1;
+
+                        @Override
+                        public FileVisitResult preVisitDirectory(
+                                @NotNull final Path dir,
+                                @NotNull final BasicFileAttributes attrs) {
+                            final String dirName = dir.getFileName().toString();
+                            // Skip dot directories
+                            if (!dirName.isEmpty() && dirName.charAt(0) == '.') {
+                                return FileVisitResult.SKIP_SUBTREE;
+                            }
+                            if (++columnCount > 0) {
+                                // We're descending and past the root
+                                final String[] components = dirName.split("=", 2);
+                                if (components.length != 2) {
+                                    throw new TableDataException(
+                                            "Unexpected directory name format (not key=value) at " + dir);
+                                }
+                                // We use an empty set to allow duplicate partition keys across files
+                                final String columnKey =
+                                        NameValidator.legalizeColumnName(components[0], Collections.emptySet());
```

Comment:
I think the correct logic is to use a new `Set` for every parse, but accumulate legalized names into the set for all columns within a parse.